main.ts

```diff
@@ -1,150 +1,293 @@
+    title = title.replace(re, '');
+    // clean up delimiters
+    const replacements = new Map([
+      [`^${delimStr}`, ''],
+      [`${delimStr}+`, delimStr],
+      [`${delimStr}(?!\ )`, titleSeparator],
+      [`(?<!\ )${delimStr}`, ''],
+    ]);
+    for (const [key, value] of replacements) {
+      let re = new RegExp(key, 'g');
+      title = title.replace(re, value);
+    }
+    return title
+  };
+
+  private readonly handleRename = async (file: TFile, oldPath: string): Promise<void> => {
+    // console.log(`file: ${oldPath} renamed to: ${file.path}`);
+    if (file instanceof TFile && file === this.app.workspace.getActiveFile()) {
+      this.app.metadataCache.onCleanCache(() => { this.refreshTitle(file); });
+    }
+  };
+
+  private readonly handleDelete = async (file: TFile): Promise<void> => {
+    this.refreshTitle();
+  };
+
+  private readonly handleOpen = async (file: TFile): Promise<void> => {
+    if (file instanceof TFile && file === this.app.workspace.getActiveFile()) {
+      this.refreshTitle(file);
+    }
+  };
+
+  private readonly handleMetaChange = async (file: TFile): Promise<void> => {
+    if (file instanceof TFile && file === this.app.workspace.getActiveFile()) {
+      this.refreshTitle(file);
+    }
+  };
+
+  private readonly handleMetaResolve = async (file: TFile): Promise<void> => {
+    if (file instanceof TFile && file === this.app.workspace.getActiveFile()) {
+      this.refreshTitle(file);
+    }
+  };
+
+  async loadSettings() {
+    this.settings = Object.assign({}, DEFAULT_SETTINGS, await this.loadData());
+  }
+
+  async saveSettings() {
+    await this.saveData(this.settings);
+  }
+
+}
+
+interface ActiveNoteTitlePluginSettings {
+  titleTemplate: string,
+  titleTemplateEmpty: string,
+  titleSeparator: string,
+  delimStr: string
+}
+
+const DEFAULT_SETTINGS: ActiveNoteTitlePluginSettings = {
+  titleTemplate: "{{basename}}~~{{vault}} - Obsidian v{{version}}",
+  titleTemplateEmpty: "{{vault}} - Obsidian v{{version}}",
+  titleSeparator: " - ",
+  delimStr: "~~"
+}
+
+class ActiveNoteTitlePluginSettingsTab extends PluginSettingTab {
+
+  plugin: ActiveNoteTitlePlugin;
+
+  constructor(app: App, plugin: ActiveNoteTitlePlugin) {
+    super(app, plugin);
+    this.plugin = plugin;
+  }
+
+  display(): void {
+    let { containerEl } = this;
+    let desc: DocumentFragment;
+    containerEl.empty();
+    containerEl.createEl('h2', {text: 'Window title templates'});
+    containerEl.createEl('p', {text: 'These two templates override the window title of the Obsidian window. This is useful for example when you use tracking software that works with window titles. You can use the format `~~{{placeholder}}~~` if you want the placeholder to be completely omitted when blank, otherwise whitespace and other characters will be preserved. You can surround a placeholder with parentheses e.g. `({{frontmatter.project}})` and it will be hidden if the referenced field is empty.'});
```

Comment:
To me it's confusing to have both `~~{{key}}~~` and `({{key}})`. I don't find any difference but could only get the last format to work properly.

Reply:
when processing the placeholder substitutions, `~~` will be replaced with the delimiter (usually ` - `). If you just put the placeholder by itself, it will be substituted but the ` - ` will not be inserted between the texts.

Maybe an example would help?

### example 1

with template string as
```
{{basename}}~~{{frontmatter.client}} ({{frontmatter.project}})~~Obsidian v{{version}}
```
and note filename: `My Fancy Note`
note contents:
```md
---
client: My Best Client
project: Very important project
---

Hello World.
```

Resulting title: `My Fancy Note - My Best Client (Very important project) - Obsidian v0.12.13`

The delimiters `~~` (or `##` , `__` if you choose in settings) only need to be one one side of the placeholder. You don't need to (and should not in fact) put them at the very start or very end of the template string.

### example 2

With template as this instead (no `~~`s)
```
{{basename}} {{frontmatter.client}} ({{frontmatter.project}}) Obsidian v{{version}}
```
(same note content)

Title would end up as `My Fancy Note My Best Client (Very important project) Obsidian v0.12.13`